### ISSUE TITLE
incorrect resolution of type variables can cause stack overflows

### ISSUE THREAD

**Issue by hishamhm** [hishamhm/tl#53]
The scope of type variables is currently not correctly controlled. This can manifest as producing accidentally recursive types — the visible symptom is a stack overflow when attempting to print the type using `show_type`. (The problem is not with `show_type`, but with the construction of the `Type` object instead.)

Here is a minimal test case, triggered by type inference on `{}`:

```
local t = {}
for i, a in ipairs(t) do
   for j, b in ipairs(a) do
      print(i, j, "value: " .. b)
   end
end
```

My plan for fixing this is via a revamp of type variables, storing them organized by scope in the main symbol table `st` alongside variables and types, instead of their own `typevars` tables.

**Comment by mwild1** [comment]
Hmm, I ran the fixed tl against the original code, and it still produces the error. This test case still fails:

```lua
local methods = {};

function methods:method1()
        return { data = function () return 1, 2, 3 end }
end

function methods:method2()
        local one, two, three
        one, two, three = self:method1():data();
        return one, two, three
end

local a = setmetatable({}, { __index = methods });

print(a:method2())
```

**Comment by hishamhm** [comment]
Fixed as part of the union types work in #63!
